lte/gateway/c/session_manager/SessionStore.cpp

```diff
@@ -122,6 +121,31 @@ bool SessionStore::update_sessions(const SessionUpdate& update_criteria) {
   return store_client_->write_sessions(std::move(session_map));
 }
 
+optional<SessionVector::iterator> SessionStore::find_session(
+    SessionMap& session_map, SessionSearchCriteria criteria) {
+  auto it = session_map.find(criteria.imsi);
+  if (it == session_map.end()) {
+    return {};
+  }
+  auto& sessions = it->second;
+  for (auto it = sessions.begin(); it != sessions.end(); ++it) {
+    switch (criteria.search_type) {
+      case IMSI_AND_SESSION_ID:
+        if ((*it)->get_session_id() == criteria.secondary_key) {
+          return it;
+        }
+      case IMSI_AND_APN:
+        if ((*it)->get_config().common_context.apn() ==
+            criteria.secondary_key) {
+          return it;
+        }
+      default:
+        continue;
+    }
+  }
+  return {};
+}
+
```

Comment:
Why do you need to return and optional? Can you not return an empty vector?

BTW, this is what I found about returning vectors. Per what i can read Cpp11 has lots of optimization to avoid copying the vector when returning 

https://stackoverflow.com/questions/15704565/efficient-way-to-return-a-stdvector-in-c

Reply:
So maybe this is my lack of C++ knowledge, but I am trying to return an iterator pointer to the session state. So it points to 1 session. I suppose I could also return a vector, but two vectors cannot share the unique pointer, it seems.

Reply:
Yes, i am on the same spot here :D

So when you say "two vectors cannot share the unique pointer", what about returning by value and let the caller to use a reference to it? Per what i read on that link, cpp should optimize that return and not copy the object twice.🙇‍♂️

Reply:
Ok, so not sure what you mean by return by value... If you mean return 'SessionState' by value, that doesn't work because SessionState can be modified by the caller. 
But maybe that is not what you mean :). By two vectors not sharing the unique pointer, the compiler literally won't let me share a unique_pointer unless I do std::move to move the owner. (Which is not good, since SessionMap should be the owner of all SessionStates.) 

Maybe what you're saying is just return `std::vector<std::unique_ptr<SessionState>&>`? But since we're always returning one SessionState anyways, I thought it would be good enough to just return the iterator. 
Not sure if that makes sense haha

Reply:
Maybe if we start finding cases where we want to return a list of session states that meet a certain criteria, then maybe we can switch to vector.

Reply:
I'm going to land this for now, since I want to put up submit reviews for stacked changes, but I am open to improving the structure. :)